src/lane.cpp

```diff
@@ -1,28 +1,65 @@
 // Header
 #include "lane.hpp"
 
-// stlib
-#include <vector>
-#include <algorithm>
-
-// Releases all graphics resources
-bool Lane::init(float world_scale)
+Lane::Lane(float world_scale, direction dir)
 {
-	m_cars = new Car[MaxCarsPerLane];
-	return true;
+	m_world_scale = world_scale;
+	m_dir = dir;
+	m_time_remaining = m_max_time_per_car;
 }
 
-void Lane::destroy()
+Lane::~Lane()
 {
-	delete m_cars;
+	m_cars.clear();
 }
 
 float Lane::get_time_remaining() const
 {
     return m_time_remaining;
 }
 
-Car* Lane::get_cars() const
+std::vector<Car> Lane::get_cars() const
 {
     return m_cars;
 }
+
+bool Lane::update(float ms)
+{
+	m_time_remaining -= ms;
+	if (m_time_remaining <= 0)
+	{
+		this->turn_car();
+		m_time_remaining = m_max_time_per_car;
+	}
+	return true;
+}
+
+void Lane::add_car(carType type)
+{
+	if (m_cars.size() < MaxCarsPerLane) {
+		// change following code based on carType once we have more than one
+		Car new_car;
+		if(new_car.init(m_world_scale)){
+			m_cars.emplace_back(new_car);
+			//new_car.enter_lane(direction dir); <-- function to animate moving car new up to previous car in line
+		}
+	}
+}
+
+void Lane::turn_car()
+{
+	if (m_cars.size() > 0) {
+		//m_cars[0].turn(); //tell the car at the front of the lane to turn
+		//wait for first car to finish turning...
+		m_cars.erase(m_cars.begin()); //Will this work? have to be careful that it won't delete the car mid-turn. Might need to like pass this as a callback or something...

```

Comment:
Maybe we could handle this in the car.turn() function? One way would be to have an erase_first_car function in the lane class that we call from the car class after the turn has been done. The upside is that this will ensure we don't run into async hell, but it does add coupling. Seems worth it to me.

Reply:
Not sure how C++ey callbacks are but that would also work

Reply:
I was thinking about using a callback as well, but I wasn't sure how to do it in C++. I think the idea about an erase_first_car function is a good idea.

Reply:
@JordanRoth 

ok, this is actually pretty hard. I wanted to have the Car class call an `erase_first_car` function from the Lane class, but I can't `#include lane.hpp` in `car.hpp` because then I'll have a circular dependency and it won't build. This also prevents me from passing a function pointer to `erase_first_car` because that also requires the Lane declaration because it's a non-static function. I don't really know how to do this besides just waiting a few seconds and calling a function in the Car class (from the Lane class) to see if it's done turning, but I think that will just cause big problems once we have collisions. Any ideas?